### PR TITLE
WT-7759 Reduce number of calls to update obsolete function

### PR DESCRIPTION
**Pull request by jiechenbo**
This ticket originally investigates the regression that was occuring in genny workloads. This change attempts to improve the regression, by restricting the number of calls made to the obsolete call functions, by grabbing the minimum timestamp within the update list, and using that as the more restrict limit. 

Previously we would use the pinned timestamp and whenever we decide to call the obsolete function again, it would check the global pinned timestamp within the __wt_txn_visible_all function. This means that this predicate would always hold true.

### CODE REVIEW

File: src/btree/row_modify.c [hunk 1]
```diff
@@ -347,6 +349,7 @@ __wt_update_obsolete_check(
                 first = upd;
         } else {
             first = NULL;
+            min_ts = upd->durable_ts;
```
Comment (hbokhari) [inline src/btree/row_modify.c:352]: Although we expect the update to be monotonically increasing in terms of commit timestamp, we still support committing out-of-order update. Can we make this check something like this:
`min_ts = WT_MIN(upd->durable_ts, min_ts)`

Comment (hbokhari) [reply src/btree/row_modify.c:352]: I am assuming the intent here is to taking the minimum of all the durable timestamps on the update chain. I would like to see more comments here.

File: src/btree/row_modify.c [hunk 2]
```diff
@@ -399,7 +402,7 @@ __wt_update_obsolete_check(
     if (count > 20 && page->modify != NULL) {
         page->modify->obsolete_check_txn = txn_global->last_running;
         if (txn_global->has_pinned_timestamp)
-            page->modify->obsolete_check_timestamp = txn_global->pinned_timestamp;
+            page->modify->obsolete_check_timestamp = WT_MAX(txn_global->pinned_timestamp, min_ts);
```
Comment (hbokhari) [inline src/btree/row_modify.c:405]: Curious why we are taking taking max of the two values? More comments will be appreciated.

Comment (jiechenbo) [reply src/btree/row_modify.c:405]: The idea is that it is potentially possible that pinned timestamp is greater than the minimum timestamp that we find in the update list. This case can only happen when finding an obsolete update but we don't successfully complete this if condition `if (first != NULL && (next = first->next) != NULL && __wt_atomic_cas_ptr(&first->next, next, NULL))`. 

Otherwise if that was not the case, this means that we always taking the minimum timestamp of any update list. I wonder if we want this to become `page->modify->obsolete_check_timestamp = WT_MIN(page->modify->obsolete_check_timestamp , min_ts);`. Let me know what your thoughts are here.

Comment (jiechenbo) [reply src/btree/row_modify.c:409]: Double checked the optimisation of `page->modify->obsolete_check_timestamp = WT_MIN(page->modify->obsolete_check_timestamp , min_ts)`, and found that it might not possible to achieve this effect. This is because the system's timestamp is always increasing, and trying to find the minimum of the timestamps, will lock the `page->modify->obsolete_check_timestamp` to the original value.

File: src/btree/row_modify.c [hunk 3]
```diff
@@ -317,16 +317,18 @@ __wt_update_obsolete_check(
     WT_TXN_GLOBAL *txn_global;
     WT_UPDATE *first, *next;
     size_t size;
-    uint64_t oldest, stable;
+    uint64_t oldest, stable, min_ts;
```
Comment (hbokhari) [inline src/btree/row_modify.c:320]: Styling: variable names should be in alphabetical order.

File: src/btree/row_modify.c (hunk 2 above)
Comment (hbokhari) [reply src/btree/row_modify.c:409]: I believe this is correct because if `min_ts` is greater than the `pinned_timestamp`, we won't be able to mark updates obsolete until the `pinned_timestamp` becomes greater. Can you reword this as a comment for future reference?

**Review by hbokhari — CHANGES_REQUESTED** [review]

File: src/btree/row_modify.c (hunk 2 above)
Comment (jiechenbo) [reply src/btree/row_modify.c:409]: I have added the comment when finding the minimum timestamp in the update list. I believe that should be suffice. Let me know if you think there is something additional we should add here.

File: src/btree/row_modify.c (hunk 1 above)
Comment (jiechenbo) [reply src/btree/row_modify.c]: Done

File: src/btree/row_modify.c (hunk 3 above)
Comment (jiechenbo) [reply src/btree/row_modify.c]: done!

**Review by hbokhari — APPROVED** [review]
LGTM

File: src/btree/row_modify.c (hunk 2 above)
Comment (hbokhari) [reply src/btree/row_modify.c:409]: Should be okay I believe.

**Comment by raviprakashgiri29** [comment]
@jiechenbo  Are you planning to change the threshold value to force evict a page as part of this ticket?

File: dist/stat_data.py [hunk 4]
```diff
@@ -202,6 +202,7 @@ def __init__(self, name, desc, flags=''):
     CacheStat('cache_bytes_max', 'maximum bytes configured', 'no_clear,no_scale,size'),
     CacheStat('cache_bytes_other', 'bytes not belonging to page images in the cache', 'no_clear,no_scale,size'),
     CacheStat('cache_bytes_updates', 'bytes allocated for updates', 'no_clear,no_scale,size'),
+    CacheStat('cache_update_obsolete_calls', 'Number of update obsolete check calls'),
```
Comment (raviprakashgiri29) [inline dist/stat_data.py:205]: Change it to `number of update obsolete check calls` to match the other stats.

Comment (sulabhM) [inline dist/stat_data.py]: I am not sure we need the stat. Other than debugging the exact problem you are trying to solve I do not see how in a customers environment it could be useful.

File: src/btree/row_modify.c [hunk 5]
```diff
@@ -317,16 +317,18 @@ __wt_update_obsolete_check(
...
 
     next = NULL;
     page = cbt->ref->page;
     txn_global = &S2C(session)->txn_global;
 
     upd_seen = upd_unstable = 0;
+    min_ts = WT_TS_NONE;
     oldest = txn_global->has_oldest_timestamp ? txn_global->oldest_timestamp : WT_TS_NONE;
     stable = txn_global->has_stable_timestamp ? txn_global->stable_timestamp : WT_TS_NONE;
+    WT_STAT_CONN_INCR(session, cache_update_obsolete_calls);
```
Comment (sulabhM) [inline src/btree/row_modify.c:331]: As I mentioned, not sure we need this stat

**Comment by jiechenbo** [comment]
> @jiechenbo Are you planning to change the threshold value to force evict a page as part of this ticket?
hey @raviprakashgiri29, found that changing the threshold value to 500 actually introduced a %5 regression within reads, query and delete commands, and the performance didn't improve as well. I later changed to 1500, and didn't see any improvement in regression. So I believe that 1000 had the best performance in my workload.

File: dist/stat_data.py (hunk 4 above)
Comment (jiechenbo) [reply dist/stat_data.py]: done

File: dist/stat_data.py [hunk 6]
```diff
@@ -202,6 +202,7 @@ def __init__(self, name, desc, flags=''):
     CacheStat('cache_bytes_max', 'maximum bytes configured', 'no_clear,no_scale,size'),
     CacheStat('cache_bytes_other', 'bytes not belonging to page images in the cache', 'no_clear,no_scale,size'),
     CacheStat('cache_bytes_updates', 'bytes allocated for updates', 'no_clear,no_scale,size'),
+    CacheStat('cache_update_obsolete_calls', 'number of update obsolete check calls'),
```
Comment (sulabhM) [inline dist/stat_data.py:205]: Not sure we need the stat here. Other than debugging this exact issue, I doubt if it will be useful in the long run. Maintaining stats could get expensive.

**Review by sulabhM — CHANGES_REQUESTED** [review]
Let us have a discussion on needing that statistic, other wise this change looks good.

File: dist/stat_data.py (hunk 4 above)
Comment (jiechenbo) [reply dist/stat_data.py]: Yeah, I wasn't sure either, so I added the statistic to make sure. I can definitely remove it.

**Review by raviprakashgiri29 — APPROVED** [review]
Changes LGTM

**Comment by jiechenbo** [comment]
Closing this PR as the effectiveness of the change doesn't seem to be much of an improvement.
